src/lib/certmap/sss_cert_digest_crypto.c

```diff
@@ -0,0 +1,169 @@
+/*
+   SSSD - digest functions - OpenSSL version
+   The calls defined here should be usable outside of SSSD as well, e.g. in
+   libsss_certmap.
+
+   Copyright (C) Sumit Bose <sbose@redhat.com> 2022
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+
+#include <string.h>
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <talloc.h>
+#include <openssl/opensslv.h>
+#include <openssl/objects.h>
+#include <openssl/evp.h>
+
+#include "lib/certmap/sss_certmap_int.h"
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+#define discard_const(ptr) ((void *)((uintptr_t)(ptr)))
+#define EVP_MD_free(ctx)
+#define EVP_MD_fetch(ctx, algorithm, properties) discard_const(EVP_get_digestbyname(algorithm))
+#endif
+
+#define DATA_STEP_SIZE 30
+
+struct get_digest_data {
+    const char **list;
+    size_t size;
+    size_t idx;
+    int error;
+};
+
+static void get_digest_helper(const OBJ_NAME *name, void *arg)
+{
+    struct get_digest_data *data = (struct get_digest_data *) arg;
+    EVP_MD *md = NULL;
+
+    if (data->error != 0) {
+        return;
+    }
+
+    if (name == NULL || name->name == NULL || islower(*name->name) == 0
+        || strstr(name->name, "RSA") != NULL
+        || strstr(name->name, "rsa") != NULL) {
+        return;
+    }
+
+    md = EVP_MD_fetch(NULL, name->name, NULL);
+    if (md != NULL) {
+        data->list[data->idx] = talloc_strdup(data->list, name->name);
+        if (data->list[data->idx] == NULL) {
+            data->error = ENOMEM;
+            return;
+        } else {
+            data->idx++;
+            if (data->idx == data->size - 1) {
+                data->size += DATA_STEP_SIZE;
+                data->list = talloc_realloc(data->list, data->list,
+                                            const char *, data->size);
+                if (data->list == NULL) {
+                    data->error = ENOMEM;
+                    return;
+                }
+            }
+            data->list[data->idx] = NULL;
+        }
+    }
+    EVP_MD_free(md);
+}
+
+
+int get_digest_list(TALLOC_CTX *mem_ctx, const char ***digest_list)
+{
+    struct get_digest_data data = { 0 };
+
+    data.size = DATA_STEP_SIZE;
+    data.list = talloc_array(mem_ctx, const char *, data.size);
+    if (data.list == NULL) {
+        return ENOMEM;
+    }
+
+    OpenSSL_add_all_algorithms();
```

Comment:
Man page says this is deprecated (as well as `OpenSSL_add_all_digests()`) and `OPENSSL_init_crypto()` should be used instead.
And `man OPENSSL_init_crypto` says it's not needed for default config.

Reply:
I switched to `OPENSSL_init_crypto()`. Even if we use the default configuration it is needed here because there will be most probably no other OpenSSL before which can initialize OpenSSL.

Reply:
`OBJ_NAME_do_all` won't do initialization?

Reply:
No, this is just some kind of iterator which has no details about the underlying data, so the data must be initialized before.